### PR TITLE
refactor(macro): make code strongly typed

### PR DESCRIPTION
**Pull request by thekip**
No logic changes. Added types wherever possible and fix issues. Also added few more tests for "black zones".

### CODE REVIEW

**Comment by thekip** [comment]
@andrii-bodnar FYI

**Review by Martin005 — CHANGES_REQUESTED** [review]
Thank you for your pull request!

I have made some suggestions and [commented one question](https://github.com/lingui/js-lingui/pull/1321#discussion_r1065853488).

From the 3 empty lines and no spaces around value names in `import`'s, I believe you are not running Prettier. Please set it up and run it on this PR.

As we have merged your PR #1329 into main, please merge main into your branch as that PR made some changes to the `packages/macro/src/index.ts` file.

File: packages/macro/src/macroJs.ts [hunk 1]
```diff
@@ -21,59 +22,33 @@ export default class MacroJs {
...
+  _expressionIndex = makeCounter()
 
-  constructor({ types }, { i18nImportName }) {
+  constructor({ types }: {types: typeof babelTypes}, { i18nImportName }: { i18nImportName: string }) {
     this.types = types
     this.i18nImportName = i18nImportName
-    this._expressionIndex = makeCounter()
   }
 
   replacePathWithMessage = (
```
[the remark this answers is not in the document]
Comment (thekip) [reply packages/macro/src/macroJs.ts:32]: I carefully checked all paths and write comprehensive TS typings and this was not used. That is a part of some outdated code.  All tests related comment and ID still passing.

File: packages/macro/src/macroJsx.ts [hunk 2]
```diff
@@ -333,13 +354,13 @@ export default class MacroJSX {
...
-  tokenizeElement = (node) => {
+  tokenizeElement = (node: JSXElement): ElementToken => {
     // !!! Important: Calculate element index before traversing children.
     // That way outside elements are numbered before inner elements. (...and it looks pretty).
     const name = this.elementIndex()
-    const children = node.children
+    const children = R.flatten(node.children
       .map((child) => this.tokenizeChildren(child))
-      .filter(Boolean)
+      .filter(() => Boolean))
```
[the remark this answers is not in the document]
Comment (thekip) [reply packages/macro/src/macroJsx.ts]: Seems this is not covered by tests

**Comment by thekip** [comment]
messed up things with rebase, looking why commits from main is showed as my

**Comment by thekip** [comment]
Done

**Review by Martin005 — APPROVED** [review]

**Comment by andrii-bodnar** [comment]
@thekip @Martin005 is this PR ready to merge?

**Comment by thekip** [comment]
Nothing to add from my side

**Pull request merged**